translator/source_code/cotrip_translator.py

```diff
@@ -0,0 +1,421 @@
+
+    # beginning_cross_street
+    properties['ending_cross_street'] = ""
+
+    # event status
+    properties['event_status'] = get_event_status(alert['alert:ReportedTime'], alert.get('alert:ExpectedEndTime'))
+
+    # type_of_work
+    # maintenance, minor-road-defect-repair, roadside-work, overhead-work, below-road-work, barrier-work, surface-work, painting, roadway-relocation, roadway-creation
+    properties['types_of_work'] = []
+
+    # reduced speed limit
+    properties['reduced_speed_limit'] = 25
+
+    # workers present
+    properties['workers_present'] = False
+
+    # restrictions
+    properties['restrictions'] = [] #will work with it late
+
+    # description
+    properties['description'] = alert['alert:Description']
+
+    # creation_date
+    properties['creation_date'] = reformat_datetime(alert['alert:ReportedTime'])
+
+    # update_date
+    
+    properties['update_date'] = reformat_datetime(alert['alert:LastUpdatedDate'])
+
+    feature = {}
+    feature['type'] = "Feature"
+    feature['properties'] = properties
+    feature['geometry'] = geometry
+
+    return feature
+
+#function to validate the alert
+def validate_alert(alert):
+    
+    if not alert or (type(alert) != dict and type(alert) != OrderedDict):
+        logging.warning('alert is empty or has invalid type')
+        return False
+    
+
+    polyline = alert.get('alert:Polyline')
+    coords = parse_polyline(polyline)
+    street = alert.get('alert:RoadName', '')
+
+    starttime = alert.get('alert:ReportedTime')
+    endtime = alert.get('alert:ExpectedEndTime')
+    description = alert.get('alert:Description')
+   
+    required_fields = [ polyline, coords, street, starttime, endtime, description]
+    for field in required_fields:
+        if not field:
+            logging.warning('Invalid alert with id =. Not all required fields are present')
+            return False
+            
+    try:
+        dateutil.parser.parse(alert['alert:ReportedTime'])
+        if alert.get('alert:ExpectedEndTime'):
+            dateutil.parser.parse(alert['alert:ExpectedEndTime'])
+    except ValueError:
+        logging.warning('Invalid alert with id . Invalid datetime format')
+        return False
+    
+    return True
+
+def reformat_datetime(datetime_string):
+    time = dateutil.parser.parse(datetime_string)
+    wzdx_format_datetime = time.astimezone(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")
+    return wzdx_format_datetime
+
+# Add ids to message
+#### This function may fail if some optional fields are not present (lanes, types_of_work, relationship, ...)
+def add_ids(message, add_ids):
+    if add_ids:
+        feed_info_id = message['road_event_feed_info']['feed_info_id']
+        data_source_id = message['road_event_feed_info']['data_sources'][0]['data_source_id']
+
+        road_event_length = len(message['features'])
+        road_event_ids = []
+        for i in range(road_event_length):
+            road_event_ids.append(str(uuid.uuid4()))
```

Comment:
Are we just using this block to generate a random road event id? We could skip this and just add a random one below.